File: 2554-maximum-number-of-integers-to-choose-from-a-range-1/main.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

class Solution {
public:
    int maxCount(vector<int>& banned, int n, int maxSum) {
        map<int, bool> bannedMap;
        for (int v : banned) {
            if (v < n) {
                bannedMap[v] = true;
            }
        }

        int totalSum = 0;
        int total = 0;

        for (int i = 1; i <= n; i++){
            if (bannedMap.find(i) != bannedMap.end()) {
                continue;
            }
            totalSum += i;
            if (totalSum > maxSum){
                return total;
            }
            total++;
        }
        return total;
    }
};

int main() {
    // Define test cases as a struct
    struct TestCase {
        int id;
        vector<int> banned;
        int n;
        int maxSum;
        int want;
    };

    // Initialize test cases
    vector<TestCase> cases = {
        {1, {1, 6, 5}, 5, 6, 2},
        {2, {1, 2, 3, 4, 5, 6, 7}, 8, 1, 0},
        {3, {11}, 7, 50, 7}
    };

    Solution solution;

    // Iterate over test cases
    for (auto& test : cases) {
        int res = solution.maxCount(test.banned, test.n, test.maxSum);
        if (res != test.want) {
            cout << "Test " << test.id << " failed: expected " << test.want 
                 << ", got " << res << endl;
        } else {
            cout << "Test " << test.id << " passed." << endl;
        }
    }

    return 0;
}

File: 2554-maximum-number-of-integers-to-choose-from-a-range-1/main.go
package main

import "fmt"

func main() {
	cases := []struct {
		id     int
		banned []int
		n      int
		maxSum int
		want   int
	}{
		{
			id:     1,
			banned: []int{1, 6, 5},
			n:      5,
			maxSum: 6,
			want:   2,
		},
		{
			id:     2,
			banned: []int{1, 2, 3, 4, 5, 6, 7},
			n:      8,
			maxSum: 1,
			want:   0,
		},
		{
			id:     3,
			banned: []int{11},
			n:      7,
			maxSum: 50,
			want:   7,
		},
	}

	for _, c := range cases {
		if res := maxCount(c.banned, c.n, c.maxSum); res != c.want {
			fmt.Printf("Test %d failed, expected %d, got %d\n", c.id, c.want, res)
		}
	}
}

func maxCount(banned []int, n int, maxSum int) int {
	m := map[int]bool{}
	for _, v := range banned {
		if v > n {
			continue
		}
		m[v] = true
	}

	totalSum := 0
	total := 0

	for i := 1; i <= n; i++ {
		if _, ok := m[i]; ok {
			continue
		}
		totalSum += i
		if totalSum > maxSum {
			return total
		}
		total++
	}

	return total
}
